### PR TITLE
Master improvement routes rules dat

### PR DESCRIPTION
**Pull request by dat-odoo**
Task link : https://www.odoo.com/web#id=1852969&action=333&active_id=131&model=project.task&view_type=form&menu_id=4720
Pad link : https://pad.odoo.com/p/r.7150665f128db847bf4d056d3d889fd3

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock/models/procurement.py [hunk 1]
```diff
@@ -35,34 +35,98 @@ class ProcurementRule(models.Model):
...
+    def _compute_action_message(self):
+        for rule in self.filtered(lambda rule: rule.location_id and rule.action and rule.picking_type_id):
+            message_dict = rule._get_message_dict()
+            message = message_dict.get((rule.action, bool(rule.location_src_id), bool(rule.location_id))) and message_dict[(rule.action, bool(rule.location_src_id), bool(rule.location_id))] or ""
+            if rule.action == 'pull_push' and rule.location_src_id and rule.location_id:
+                message = message_dict[('pull', bool(rule.location_src_id), bool(rule.location_id))] + "<br/><br/>"\
+                    + message_dict[('push', bool(rule.location_src_id), bool(rule.location_id))]
+            rule.rule_message = message
+
+    def _apply(self, move):
```
Comment (pimodoo) [inline addons/stock/models/procurement.py]: - Should be multi
- Should check the action_type
- move push _apply from stock_move here, with api.model

File: addons/stock/models/procurement.py [hunk 2]
```diff
@@ -83,7 +147,7 @@ def _run_move(self, product_id, product_qty, product_uom, location_id, name, ori
 
     def _get_stock_move_values(self, product_id, product_qty, product_uom, location_id, name, origin, values, group_id):
         ''' Returns a dictionary of values that will be used to create a stock move from a procurement.
-        This function assumes that the given procurement has a rule (action == 'move') set on it.
+        This function assumes that the given procurement has a rule (action == 'pull') set on it.
```
Comment (pimodoo) [inline addons/stock/models/procurement.py]: pull or pull_push

File: addons/stock/models/procurement.py [hunk 3]
```diff
@@ -179,14 +243,14 @@ def run(self, product_id, product_qty, product_uom, location_id, name, origin, v
         values.setdefault('priority', '1')
         values.setdefault('date_planned', fields.Datetime.now())
         rule = self._get_rule(product_id, location_id, values)
-
+        action = 'pull' if rule.action == 'pull_push' else rule.action
```
Comment (pimodoo) [inline addons/stock/models/procurement.py]: ->248

File: addons/stock/models/procurement.py [hunk 4]
```diff
@@ -179,14 +243,14 @@ def run(self, product_id, product_qty, product_uom, location_id, name, origin, v
         values.setdefault('priority', '1')
         values.setdefault('date_planned', fields.Datetime.now())
         rule = self._get_rule(product_id, location_id, values)
-
```
Comment (pimodoo) [inline addons/stock/models/procurement.py]: - think about a solution to call run(push) or run(pull)
- look at search_rule and push_apply duplication

**Comment by amoyaux** [comment]
Don't remove the delay on push_rule in test_02_product_route_level_delays
It seems that it is not possible to specify a different delay between push and pull any more now. Instead of removing the delay on push_pull adapt the date in assert it would be more interesting to check that delay on pull works.

File: addons/stock/models/procurement.py (hunk 1 above)
Comment (dat-odoo) [reply addons/stock/models/procurement.py]: one rule can be applied to it should a single one 
no need to check action as we already filtered rule based on (push, pull_push)

For third point we have to change all mechanism of applying push rule.... ( is it ok )

File: addons/stock/models/procurement.py (hunk 2 above)
Comment (dat-odoo) [reply addons/stock/models/procurement.py]: ok

File: addons/stock/models/procurement.py [hunk 5]
```diff
@@ -35,34 +35,98 @@ class ProcurementRule(models.Model):
...
+    warehouse_id = fields.Many2one('stock.warehouse', 'Warehouse')
     propagate_warehouse_id = fields.Many2one(
         'stock.warehouse', 'Warehouse to Propagate',
         help="The warehouse to propagate on the created move/procurement, which can be different of the warehouse this rule is for (e.g for resupplying rules from another warehouse)")
-
-    def _run_move(self, product_id, product_qty, product_uom, location_id, name, origin, values):
+    auto = fields.Selection([
+        ('manual', 'Manual Operation'),
+        ('transparent', 'Automatic No Step Added')], string='Automatic Move',
+        default='manual', index=True, required=True,
```
[the remark this answers is not in the document]
Comment (dat-odoo) [reply addons/stock/models/procurement.py]: This action done by user manually as auto Apply only on push rule not for pull.

**Comment by dat-odoo** [comment]
@amoyaux , old test case was wrong ( As there is nothing related to push delay ) 
but we can add new test case to make it better.

but in pull_push it will apply same delay on both ,if some one need different delay they should create pull and push separate.

File: addons/stock/models/procurement.py (hunk 4 above)
Comment (jir-odoo) [reply addons/stock/models/procurement.py]: No mechanism for grouping , propagate warehouse for push rule before , are you want to implement it ? I prefer to make new one with business cases

File: addons/purchase_stock/models/stock_rule.py [hunk 6]
```diff
@@ -16,14 +16,17 @@ class StockRule(models.Model):
     def _get_message_dict(self):
         message_dict = super(StockRule, self)._get_message_dict()
         message_dict.update({
-            ('buy', False, True): _('When products are needed in %s, <br/> a request for quotation is created to fulfill the need.') % (self.location_id.display_name,)
+            ('buy', False, True): _('When products are needed in <b><i>%s</i></b>, <br/> a request for quotation is created to fulfill the need.') % (self.location_id.display_name,)
```
Comment (sle-odoo) [inline addons/purchase_stock/models/stock_rule.py]: @dat-odoo shouldn't you adapt `_search_rule` to handle the global rules?
also please refer to rev https://github.com/odoo/odoo/commit/b1b3de1339a13504ac5e223aec465bd70417d11e#diff-23b5396fd047e334041d578a52ace527 in your commit message

File: addons/stock/data/stock_data.xml [hunk 7]
```diff
@@ -80,13 +80,13 @@
...
         <!--
-             Procurement rules
+             Stock rules
         -->
         
         <record id="route_warehouse0_mto" model='stock.location.route'>
             <field name="name">Make To Order</field>
             <field name="company_id"></field>
-            <field name="sequence">10</field>
+            <field name="sequence">5</field>
```
Comment (sle-odoo) [inline addons/stock/data/stock_data.xml]:
```
<!-- set a lower sequence on the mto route than on the resupply routes -->
```

File: addons/stock/models/stock_move.py [hunk 8]
```diff
@@ -505,31 +504,17 @@ def _do_unreserve(self):
         return True
 
     def _push_apply(self):
-        # TDE CLEANME: I am quite sure I already saw this code somewhere ... in routing ??
-        Push = self.env['stock.location.path']
         for move in self:
             # if the move is already chained, there is no need to check push rules
-            if move.move_dest_ids:
+            if move.move_dest_ids or move.procure_method == 'make_to_order':
```
Comment (sle-odoo) [inline addons/stock/models/stock_move.py]: seems weird
the last move of a buy + pick pack ship is mto and we should be apply to call _push_apply

File: addons/stock/models/stock_move.py [hunk 9]
```diff
@@ -505,31 +504,17 @@ def _do_unreserve(self):
...
-                if product_routes:
-                    rules = Push.search(expression.AND([[('route_id', 'in', product_routes.ids)], domain]), order='route_sequence, sequence', limit=1)
-            if not rules:
-                # TDE FIXME/ should those really be in a if / elif ??
-                # then we search on the warehouse if a rule can apply
-                if move.warehouse_id:
-                    rules = Push.search(expression.AND([[('route_id', 'in', move.warehouse_id.route_ids.ids)], domain]), order='route_sequence, sequence', limit=1)
-                elif move.picking_id.picking_type_id.warehouse_id:
-                    rules = Push.search(expression.AND([[('route_id', 'in', move.picking_id.picking_type_id.warehouse_id.route_ids.ids)], domain]), order='route_sequence, sequence', limit=1)
+            warehouse_id = move.warehouse_id and move.warehouse_id or move.picking_id.picking_type_id.warehouse_id
```
Comment (sle-odoo) [inline addons/stock/models/stock_move.py]: simply move.warehouse_id or move.picking_id.picking_type_id.warehouse_id should work?

File: addons/stock/models/stock_move.py [hunk 10]
```diff
@@ -505,31 +504,17 @@ def _do_unreserve(self):
...
+            rules = self.env['procurement.group']._search_rule(move.route_ids, move.product_id, warehouse_id, domain)
```
Comment (sle-odoo) [inline addons/stock/models/stock_move.py]: what if warehouse_id is empty, can you double check that _search_rule correctly skip the search?

File: addons/stock/models/stock_rule.py [hunk 11]
```diff
@@ -35,36 +35,143 @@ class ProcurementRule(models.Model):
...
-    location_src_id = fields.Many2one('stock.location', 'Source Location', help="Source location is action=move")
+    company_id = fields.Many2one('res.company', 'Company',
+        default=lambda self: self.env.user.company_id)
+    location_id = fields.Many2one('stock.location', 'Destination Location', required=True)
+    location_src_id = fields.Many2one('stock.location', 'Source Location')
     route_id = fields.Many2one('stock.location.route', 'Route', required=True, ondelete='cascade')
     procure_method = fields.Selection([
         ('make_to_stock', 'Take From Stock'),
-        ('make_to_order', 'Create Procurement')], string='Move Supply Method',
+        ('make_to_order', 'Trigger Another Rule')], string='Move Supply Method',
```
Comment (sle-odoo) [inline addons/stock/models/stock_rule.py]: is it possible to set a push rule mto? shouldn't we block this?

File: addons/stock/models/stock_rule.py [hunk 12]
```diff
@@ -35,36 +35,143 @@ class ProcurementRule(models.Model):
...
     propagate = fields.Boolean(
         'Propagate cancel and split', default=True,
-        help='If checked, when the previous move of the move (which was generated by a next procurement) is cancelled or split, the move generated by this move will too')
-    warehouse_id = fields.Many2one('stock.warehouse', 'Served Warehouse', help='The warehouse this rule is for')
+        help="When ticked, if the move is splitted or cancelled, the next move will be too.")
+    warehouse_id = fields.Many2one('stock.warehouse', 'Warehouse')
     propagate_warehouse_id = fields.Many2one(
         'stock.warehouse', 'Warehouse to Propagate',
         help="The warehouse to propagate on the created move/procurement, which can be different of the warehouse this rule is for (e.g for resupplying rules from another warehouse)")
+    auto = fields.Selection([
```
Comment (sle-odoo) [inline addons/stock/models/stock_rule.py]: is it possible to set auto on a pull rule? shouldn't we block this?

File: addons/stock/models/stock_rule.py [hunk 13]
```diff
@@ -35,36 +35,143 @@ class ProcurementRule(models.Model):
...
+        """
+        self.location_src_id = self.picking_type_id.default_location_src_id.id
+        self.location_id = self.picking_type_id.default_location_dest_id.id
 
-    def _run_move(self, product_id, product_qty, product_uom, location_id, name, origin, values):
+    @api.onchange('route_id', 'company_id')
+    def _onchange_route(self):
+        """ If a route is set and it exists a route company, ensure that the
+        rule's company is the same than the route's company.
+        """
```
Comment (sle-odoo) [inline addons/stock/models/stock_rule.py]: now that we removed the global rules, the first part of the docstring is outdated

File: addons/stock/models/stock_rule.py [hunk 14]
```diff
@@ -35,36 +35,143 @@ class ProcurementRule(models.Model):
...
+        company_id = self.route_id.company_id
+        if company_id:
+            self.company_id = company_id
+            if self.picking_type_id.warehouse_id.company_id != company_id:
+                self.picking_type_id = False
+        domain = {'company_id': company_id and [('id', '=', company_id.id)] or []}
+        return {'domain': domain}
+
+    def _get_message_values(self):
+        """ Return the source, destination and picking_type apply on a stock
```
Comment (sle-odoo) [inline addons/stock/models/stock_rule.py]: applied

File: addons/stock/models/stock_rule.py [hunk 15]
```diff
@@ -35,36 +35,143 @@ class ProcurementRule(models.Model):
...
+        rule. The purpose of this function is to avoid code duplication in
+        _get_message_dict functions since it often require those data.
```
Comment (sle-odoo) [inline addons/stock/models/stock_rule.py]: requires

File: addons/stock/models/stock_rule.py [hunk 16]
```diff
@@ -35,36 +35,143 @@ class ProcurementRule(models.Model):
...
+        """
+        for rule in self.filtered(lambda rule: rule.action):
+            message_dict = rule._get_message_dict()
+            message = message_dict.get(rule.action) and message_dict[rule.action] or ""
+            if rule.action == 'pull_push':
+                message = message_dict['pull'] + "<br/><br/>" + message_dict['push']
+            rule.rule_message = message
+
+    def _apply(self, move):
+        """ Apply a rule on a move.
```
Comment (sle-odoo) [inline addons/stock/models/stock_rule.py]: apply a push rule

File: addons/stock/models/stock_rule.py [hunk 17]
```diff
@@ -35,36 +35,143 @@ class ProcurementRule(models.Model):
...
+                move._push_apply()
+        else:
+            new_move_vals = self._prepare_move_copy_values(move, new_date)
+            new_move = move.copy(new_move_vals)
+            move.write({'move_dest_ids': [(4, new_move.id)]})
+            new_move._action_confirm()
+
+    def _prepare_move_copy_values(self, move_to_copy, new_date):
+        new_move_vals = {
+                'origin': move_to_copy.origin or move_to_copy.picking_id.name or "/",
```
Comment (sle-odoo) [inline addons/stock/models/stock_rule.py]: wrong indent

File: addons/stock/models/stock_rule.py (hunk 16 above)
Comment (sle-odoo) [reply addons/stock/models/stock_rule.py]: shoudln't we prefix all methods related to push rule with _push, same with _pull?

Comment (sle-odoo) [reply addons/stock/models/stock_rule.py]: _run_push

File: addons/stock/models/stock_rule.py (hunk 17 above)
Comment (sle-odoo) [reply addons/stock/models/stock_rule.py]: prefix this method with something related to push

File: addons/stock/models/stock_rule.py [hunk 18]
```diff
@@ -179,43 +286,43 @@ def run(self, product_id, product_qty, product_uom, location_id, name, origin, v
         values.setdefault('priority', '1')
         values.setdefault('date_planned', fields.Datetime.now())
         rule = self._get_rule(product_id, location_id, values)
```
Comment (sle-odoo) [inline addons/stock/models/stock_rule.py]: maybe add a comment about _get_rule returing all rules except the push ones

File: addons/stock/models/stock_warehouse.py [hunk 19]
```diff
@@ -323,37 +320,27 @@ def _create_or_update_delivery_route(self, routes_data):
             if warehouse.delivery_route_id:
                 delivery_route = warehouse.delivery_route_id
                 delivery_route.write({'name': warehouse._format_routename(route_type=warehouse.delivery_steps)})
-                delivery_route.pull_ids.write({'active': False})
+                delivery_route.rule_ids.write({'active': False})
```
Comment (sle-odoo) [inline addons/stock/models/stock_warehouse.py]: filter on pull
what about pull_push? :)

File: addons/stock/models/stock_warehouse.py [hunk 20]
```diff
@@ -566,31 +545,33 @@ def _update_reception_delivery_resupply(self, reception_new, delivery_new):
...
 
     def _check_reception_resupply(self, new_location):
         """ Check routes being delivered by the warehouses (resupply routes) and
         change their rule coming from the transit location """
         routes = self.env['stock.location.route'].search([('supplied_wh_id', 'in', self.ids)])
-        self.env['procurement.rule'].search([
-            '&', ('route_id', 'in', routes.ids),
+        self.env['stock.rule'].search([
+            '&', '&', ('route_id', 'in', routes.ids),
+            ('action', '!=', 'push'),
```
Comment (sle-odoo) [inline addons/stock/models/stock_warehouse.py]: maybe reindent for clarity sake

File: addons/stock/views/stock_rule_views.xml [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+                            <group>
+                                <field name="action"/>
+                                <field name="picking_type_id"/>
+                                <field name="location_src_id" attrs="{'required': [('action', 'in', ['pull', 'push', 'pull_push'])]}"/>
+                                <field name="location_id"/>
+                                <field name="auto" attrs="{'invisible': [('action', 'not in', ['push', 'pull_push'])]}"/>
+                                <field name="procure_method" attrs="{'invisible': [('action', 'not in', ['pull', 'pull_push'])]}"/>
+                            </group>
+                            <group>
+                                <div colspan ="2">
```
Comment (sle-odoo) [inline addons/stock/views/stock_rule_views.xml]: unnecessary whitespace

File: addons/stock/views/stock_rule_views.xml [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+                                <field name="partner_address_id" attrs="{'invisible': [('action', '=', 'push')]}"/>
+                                <field name="delay"/>
+                            </group>
+                        </group>
+                    </sheet>
+                </form>
+            </field>
+        </record>
+
+        # Form view for route
```
Comment (sle-odoo) [inline addons/stock/views/stock_rule_views.xml]: wrong comment formatting

File: addons/stock/models/stock_move.py (hunk 10 above)
Comment (amoyaux) [reply addons/stock/models/stock_move.py]: if warehouse_routes:
                res = Rule.search(expression.AND([[('route_id', 'in', warehouse_routes.ids)], domain]), order='route_sequence, sequence', limit=1)
        return res

File: addons/stock/models/stock_rule.py (hunk 11 above)
Comment (amoyaux) [reply addons/stock/models/stock_rule.py]: Hide in view

File: addons/stock/models/stock_rule.py (hunk 12 above)
Comment (amoyaux) [reply addons/stock/models/stock_rule.py]: Also hide in view

File: addons/stock_dropshipping/data/stock_data.xml [hunk 23]
```diff
@@ -19,16 +19,16 @@
...
             <field name="company_id"></field>
             <field name="sale_selectable" eval="True"/>
             <field name="product_selectable" eval="True"/>
             <field name="product_categ_selectable" eval="True"/>
         </record>
 
-        <record id="procurement_rule_drop_shipping" model="procurement.rule">
-            <field name="name">Vendor → Customer</field>
+        <record id="stock_rule_drop_shipping" model="stock.rule">
+            <field name="name">Vendor → Customer(Dropshipping)</field>
```
Comment (sle-odoo) [inline addons/stock_dropshipping/data/stock_data.xml]: space after customer

File: addons/stock/models/stock_warehouse.py (hunk 19 above)
Comment (amoyaux) [reply addons/stock/models/stock_warehouse.py]: not filter on pull. It will archive all the rules on the route. Then modify the route name and search or create all the default rules for depends the delivery steps
